### DISCUSSION TITLE
How do I?

### DISCUSSION THREAD

**Discussion by petermr** [petermr/petermr#51] [Announcements]
# How do I do something?

These are common questions which apply to most of us, and which can be answered precisely

Please 
* add you own questions (COMMENTS)
* REPLY to others

## List of Comments
This is a simple Table Of Contents of this discussion. (Manually created)

* [How do I create a dictionary from a list of words?] ( https://github.com/petermr/petermr/discussions/51#discussioncomment-3431169)

**Comment by petermr** [comment]
# How do I create a dictionary from a list of words?

I have a list of words and want to create a dictionary (manual)
* I may or may not have WikidataIDs for them

I have a list of words in a file (we can use `pyami`)

**Reply by petermr** [reply]
# manual method

This is not the recommended method as it is easy to make mistakes but is necessary sometimes. Please use `docanalysis` or `pyami` if possible.)

(All dictionaries are XML. Always use a text editor , NOT Word.)

The simplest dictionary must have a title, which is also the stem of the filename, and has at least one `entry` with a `term` attribute.

## Example
in `emissions.xml`

```
<dictionary title="emissions" version="0.0.1">
  <entry term="carbon budget"></entry>
</dictionary>
```
We often have `desc` (description) and `wikidataID` , example:

```
<dictionary title="emissions" version="0.0.1">
  <entry term="carbon budget" wikidataID="Q16722080" desc="limit on CO2 emission for a given climate impact"> 
  </entry>
</dictionary>
```
Where possible descriptions should be taken from Wikidata.

**Comment by petermr** [comment]
# how do I use a search to create a list of words (and then a dictionary)?

`docanalysis` can search a file or corpus and create a wordlist (with frequencies), optionally using stopwords and stemming.

**Reply by ShweataNHegde** [reply]
We have a tutorial on this: https://github.com/petermr/docanalysis/wiki/docanalysis-Tutorial#create-ami-dict-of-entities

**Comment by petermr** [comment]
# How do I use Github to manage my chapter and dictionary?

**Comment by petermr** [comment]
# How can I use dictionaries to search documents?

**Comment by petermr** [comment]
# How do I create an IPCC Chapter in HTML?

**Comment by petermr** [comment]
# How do I add one or more PDFs to the CProject?

This is done through `pyami` , 'make_project` option. This can create a `CProject` from scratch or add new files to an existing one.

**Comment by petermr** [comment]
# How do I create a `CProject` and `CTrees` from a list of PDFs? 

## What's a CProject and CTree?

A "CProject" is simply the top folder of a Corpus. (Think "CorpusProject"). A corpus is a collection of documents and associated files (some original, some computed/derived). Each meaningful document generates a `CTree` folder, often with hundreds of files.  (A PDF or HTML "document" is not logically a single chunk and can have many images, tables, metadata, supplemental info and computed information. All these are held as **files** in your normal folders - there  is no hidden magic and you can use many tools (UNIX, MS, Pandas, etc. to manage them). What we require is simply:
* keep the filenames we create (e.g. `fulltext.*` is a **reserved filename**; if you rename it, `AMI` tools stop working).
* don't move the folders or files 

## creating the `CProject`
The PDFs should all be in a single directory (`/some/where` ; NOT the real name; use YOUR folder name)
```
some/
    where/
          a.pdf
          b.pdf
          c.pdf
```

To create the `CProject` run `pyami`. Until this is installed you'll either have to use the filepath of the AMI software or use the full name/s of the data file/s.

`cd` to `py4ami` modules

Run 
```
python -m py4ami/ami_project -p /some/where --make --max_flags 0
```

the `max_flags` stops old PDFs creating a new directory.

This will create:
```
some/
    where/
        a/
           fulltext.pdf
        b/
            fulltext.pdf
        c/
            fulltext.pdf
```
If you don't get this format, ask for help.

## 

## adding new files

We often need to add new 
copy new files (`d.pdf`) into `/some/where` project
Run command as above

This will create:
```
some/
    where/
        a/
            fulltext.pdf
        b/
            fulltext.pdf
        c/
            fulltext.pdf
        d/
            fulltext.pdf
```

**Comment by petermr** [comment]
# Download or install software for petermr projects

You will need:
* `git` (either commandline or tool (PyCharm, etc.)
*

Software includes:
* `pygetpapers` (available as PyPI https://pypi.org/project/pygetpapers/)
* `docanalysis` (https://pypi.org/project/docanalysis/)
* `pyami`

**Reply by petermr** [reply]
# cloning `pyami`

```
<cd to your software directory>

git clone https://github.com/petermr/pyami.git
```

**Reply by petermr** [reply]
# How do I install py4ami?

There is currently no "pip install" for `py4ami`. This is high priority. Keep nagging @petermr for an install version.

## PyPI
We now call the software `py4ami` because of name conflicts.  It is a project on `PyPI`: https://pypi.org/project/py4ami/

### `tar.gz`
Currently there is a `tar.gz` on PyPI . This may be several versions behind the development version. Keep pushing me for a `pip install` version.

### git clone
This is the best current version. You have to:
* clone the repo 
```
cd /some/where # NOT the actual name; replace by your software folder name
git clone https://github.com/petermr/pyami.git
```
This will create a folder `/some/where/pyami/` (the "local repo"). To run, your simplest approach will be to 
```
cd /some/where/pyami
```
and see below.

**Reply by petermr** [reply]
# How do I run pyami?

**This is temporary**

`py4ami` is organised as a Python project:

```
py4ami
├── CONFIG.md
├── FILES.md
├── Pyami.md
├── README.md
├── SEARCH.md
├── TODO.md
├── __init__.py
├── ami_bib.py
├── ami_chem.py
├── ami_config.py
├── ami_convert.py
├── ami_demos.py
├── ami_dict.py
├── ami_gui.py
├── ami_html.py
├── ami_pdf.py
├── ami_project.py
├── ami_sections.py
├── bbox_copy.py
├── config.ini
├── constants.py
├── examples.py
├── file_lib.py
├── fulltext.pdf.txt
├── gutil.py
├── jats_lib.py
├── pdfreader.py
├── pyami.ini
├── pyamix.py
├── resources
│   ├── SmartStoplist.txt
│   ├── amidicts
│   │   ├── VC_EthnobotanicalUse.xml
...
```

run from the directory above `pyami` (the one you have checked out)

```
cd /some/where
git clone https://github.com/petermr/pyami.git
cd pyami
```
You are now in the right directory to run `pyami`
```
 python -m py4ami.pyamix --help
```
will run the main entry point (`pyamix`) giving the original options. Some of these have been superseded by more specific commands (see below)

```
usage: pyamix.py [-h]
                 [--apply {pdf2txt,pdf2svg,svg2page,txt2para,txt2sent,xml2html,xml2sect,xml2txt} [{pdf2txt,pdf2svg,svg2page,txt2para,txt2sent,xml2html,xml2sect,xml2txt} ...]]
                 [--assert ASSERT [ASSERT ...]] [--combine COMBINE] [--config CONFIG [CONFIG ...]]
                 [--copy COPY [COPY ...]] [--debug DEBUG [DEBUG ...]] [--delete DELETE [DELETE ...]]
                 [--dict DICT [DICT ...]] [--examples [EXAMPLES [EXAMPLES ...]]] [--filter FILTER [FILTER ...]]
                 [--flags FLAGS [FLAGS ...]] [--glob GLOB [GLOB ...]] [--languages LANGUAGES [LANGUAGES ...]]
                 [--loglevel LOGLEVEL] [--maxbars [MAXBARS]] [--nosearch] [--outfile OUTFILE]
                 [--patt PATT [PATT ...]] [--plot] [--proj PROJ [PROJ ...]] [--sect SECT [SECT ...]]
                 [--split {txt2para,xml2sect} [{txt2para,xml2sect} ...]]

Search sections with dictionaries and patterns

optional arguments:
  -h, --help            show this help message and exit
  --apply {pdf2txt,pdf2svg,svg2page,txt2para,txt2sent,xml2html,xml2sect,xml2txt} [{pdf2txt,pdf2svg,svg2page,txt2para,txt2sent,xml2html,xml2sect,xml2txt} ...]
                        list of sequential transformations (1:1 map) to apply to pipeline ({self.TXT2SENT} NYI)
  --assert ASSERT [ASSERT ...]
                        assertions; failure gives error message (prototype)
  --combine COMBINE     operation to combine files into final object (e.g. concat text or CSV path
  --config CONFIG [CONFIG ...], -c CONFIG [CONFIG ...]
                        path (e.g. ~/pyami/config.ini.master) with list of config path(s) or config vars;
                        "symbols": gives symbols
  --copy COPY [COPY ...]
                        copy path or directory from=<from> to=<to> overwrite=<yes/no default=no>
  --debug DEBUG [DEBUG ...]
                        debugging commands , symbols, numbers, (not formalised)
  --delete DELETE [DELETE ...]
                      delete globbed files. Argument/s <glob> are relative to `proj`
  --dict DICT [DICT ...], -d DICT [DICT ...]
                        dictionaries to ami-search with, _help gives list
  --examples [EXAMPLES [EXAMPLES ...]]
                        simple demos, empty gives list, "all" runs all. May need downloading corpora
  --filter FILTER [FILTER ...]
                        expr to filter with
  --flags FLAGS [FLAGS ...]
                        name-value pairs collected into self.flag_dict, "help" gives list
  --glob GLOB [GLOB ...], -g GLOB [GLOB ...]
                        glob files; python syntax (* and ** wildcards supported); include alternatives in
                        {...,...}.
  --languages LANGUAGES [LANGUAGES ...]
                        languages (NYI)
  --loglevel LOGLEVEL, -l LOGLEVEL
                        log level (NYI)
  --maxbars [MAXBARS]   max bars on plot (NYI)
  --nosearch            search (NYI)
  --outfile OUTFILE     output path. default is single path (default is overwrite). expands special variables
                        _CPROJ, _CTREE, _PARENT to create iterators NYI
  --patt PATT [PATT ...]
                        patterns to search with (NYI); regex may need quoting
  --plot                plot params (NYI)
  --proj PROJ [PROJ ...], -p PROJ [PROJ ...]
                        projects to search; _help will give list
  --sect SECT [SECT ...], -s SECT [SECT ...]
                        sections to search; _help gives all(?)
  --split {txt2para,xml2sect} [{txt2para,xml2sect} ...]
                        split fulltext.* into paras, sections
```
## more specific commands are organized by subsections

Example:
```
python -m py4ami.ami_project --help
```
with commands:
```
usage: ami_project.py [-h] [--project PROJECT] [--make] [--formats FORMATS [FORMATS ...]] [--keep]
                      [--max_len MAX_LEN] [--max_flag MAX_FLAG]

Project parsing

optional arguments:
  -h, --help            show this help message and exit
  --project PROJECT     project directory
  --make                make project from list of filetypes
  --formats FORMATS [FORMATS ...]
                        input formats
  --keep                keep original PDFs
  --max_len MAX_LEN     max length of project name
  --max_flag MAX_FLAG   max number of disambiguation flags '_
```

**Comment by petermr** [comment]
# How do I run the "core" Jupyter Notebook on Google Colabs?

@ShweataNHegde , @ananya and others created the inspirational Jupyter Notebook on Google Colabs. Where is it and how do I run it?
